### PR TITLE
WIP: categorical features

### PR DESCRIPTION
**Pull request by mchernov-intel**

### CODE REVIEW

File: cpp/oneapi/dal/test/engine/feature_types.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+feature_types_builder& feature_types_builder::set(const range& r, feature_type type) {
+    impl_->set(r, type);
+    return *this;
+}
+
+feature_types feature_types_builder::build() const {
+    return impl_->build();
+}
+
+} // namespace oneapi::dal::test::engine
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/test/engine/feature_types.cpp:65]: Please add end of line

**Comment by rlnx** [comment]
@michael-smirnov, please have a look at the homogen builder

File: cpp/oneapi/dal/algo/pca/test/batch.cpp [hunk 2]
```diff
@@ -54,6 +55,20 @@ class pca_batch_test : public te::algo_fixture {
...
+        }
+        if (meta.get_feature_type(2) == feature_type::nominal) {
+            fmt::print("it works2!");
+        }
+        if (meta.get_feature_type(3) == feature_type::nominal) {
+            fmt::print("it works3!");
+        }
+        if (meta.get_feature_type(4) == feature_type::ordinal) {
+            fmt::print("it works4!");
+        }
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/pca/test/batch.cpp:70]: This looks like a silent test on table metadata features.That's not the right place for such code - if you want to have tests on metadata, create separate tests.

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 3]
```diff
@@ -75,8 +75,20 @@ class homogen_table_builder_impl {
         column_count_ = 0;
     }
 
-    void set_feature_type(feature_type ft) {
-        throw dal::unimplemented(dal::detail::error_messages::method_not_implemented());
+    void set_feature_type(feature_type ft, std::int64_t idx) {
+        if (column_count_ == 0) {
+            throw dal::domain_error{ "column_count should be > 0 " };
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:80]: Use message ids

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 4]
```diff
@@ -75,8 +75,20 @@ class homogen_table_builder_impl {
...
+        }
+        ONEDAL_ASSERT(idx >= 0);
+        ONEDAL_ASSERT(idx < column_count_);
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:83]: not sure these have to be an asserts

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 5]
```diff
@@ -75,8 +75,20 @@ class homogen_table_builder_impl {
...
+
+        allocate_ftypes_if_need();
+        feature_type* ftypes_ptr = ftypes_.need_mutable_data().get_mutable_data();
+        ftypes_ptr[idx] = ft;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:87]: `ftypes_` are not reset in `reset()` methods at least. Make sure that the build logic is not broken in other scenarios.

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 6]
```diff
@@ -227,11 +235,23 @@ class homogen_table_builder_impl {
         }
     }
 
+    void allocate_ftypes_if_need() {
+        ONEDAL_ASSERT(column_count_ != 0);
+        if (ftypes_.get_count() == 0) {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:240]: Why not `fptypes_.get_count() != column_count_`?

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 7]
```diff
@@ -227,11 +235,23 @@ class homogen_table_builder_impl {
         }
     }
 
+    void allocate_ftypes_if_need() {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:238]: Why you created a separate method to allocate `ftypes_` but do not allocating it in `allocate` method?

File: cpp/oneapi/dal/table/backend/homogen_table_impl.cpp [hunk 8]
```diff
@@ -30,6 +30,21 @@ table_metadata create_homogen_metadata(std::int64_t feature_count, data_type dty
     return table_metadata{ dtypes, ftypes };
 }
 
+table_metadata create_homogen_metadata(std::int64_t feature_count,
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_impl.cpp:33]: I'd suggest removing that function and create `dtypes` array explicitly then passing it to the `table_metadata` ctor.

File: cpp/oneapi/dal/table/backend/homogen_table_impl.hpp [hunk 9]
```diff
@@ -39,27 +42,21 @@ class homogen_table_impl {
...
-        if (layout != data_layout::row_major && layout != data_layout::column_major) {
-            throw dal::domain_error(error_msg::unsupported_data_layout());
-        }
+    homogen_table_impl(std::int64_t row_count,
+                       std::int64_t column_count,
+                       const array<byte_t>& data,
+                       const array<feature_type>& ftypes,
+                       data_type dtype,
+                       data_layout layout)
+            : meta_(create_homogen_metadata(column_count, dtype, ftypes)),
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/table/backend/homogen_table_impl.hpp:54]: What if `ftypes` element count is not equal to `column_count`?

**Pull request closed without merging**